### PR TITLE
Upgrade treetime 0.9

### PR DESCRIPTION
**Pull request by rneher**
upgrades TreeTime to version == 0.9.*

This add an argument `--use-fft` to `augur refine`. The full benefit of this will only be felt when using `--date-inference marginal`. 

Also resolves https://github.com/nextstrain/augur/issues/1019

### Checklist

- [x] Add a message in [CHANGES.md](https://github.com/nextstrain/augur/blob/HEAD/CHANGES.md) summarizing the changes in this PR. Keep headers and formatting consistent with the rest of the file.

### CODE REVIEW

**Comment by rneher** [comment]
this test fail with this message:
```
ERROR: Augur version incompatibility detected: the JSON translate/data/zika/nt_muts.json was generated by {'program': 'augur', 'version': '16.0.3'}, which is incompatible with the current augur version (17.0.0). We suggest you rerun the pipeline using the current version of augur.
   Validating schema of 'translate/data/zika/nt_muts.json'...
```
not sure how we end up with a Schroedinger cat augur installations.

**Comment by corneliusroemer** [comment]
Thanks! I'll test this and have a look at the failing test

File: setup.py [hunk 1]
```diff
@@ -58,7 +58,7 @@
         "networkx >= 2.5, ==2.*",
         "packaging >=19.2",
         "pandas >=1.0.0, ==1.*",
-        "phylo-treetime ==0.8.*",
+        "phylo-treetime ==0.9.*",
```
Comment (corneliusroemer) [inline setup.py:61]: We may want to start off with the latest bug fix version of treetime here? Why not say `>=0.9.2, ==0.9.*`?

File: augur/refine.py [hunk 2]
```diff
@@ -2,17 +2,17 @@
...
 from .io import read_metadata
 from .utils import read_tree, write_json, InvalidTreeError
-from treetime.vcf_utils import read_vcf, write_vcf
+from treetime.vcf_utils import read_vcf
 from treetime.seq_utils import profile_maps
 
 def refine(tree=None, aln=None, ref=None, dates=None, branch_length_inference='auto',
              confidence=False, resolve_polytomies=True, max_iter=2, precision='auto',
-             infer_gtr=True, Tc=0.01, reroot=None, use_marginal=False, fixed_pi=None,
+             infer_gtr=True, Tc=0.01, reroot=None, use_marginal='always', fixed_pi=None, use_fft=True,
```
Comment (corneliusroemer) [inline augur/refine.py:15]: Does this change the default argument for `use_marginal` when invoked via the CLI?

If so, we'd want to make this a major version increase for Augur.

**Comment by corneliusroemer** [comment]
The version mismatch is fixed.

One failing test remains but it seems to be unrelated - something about TB and translate.

It's the recent test added by @huddlej for https://github.com/nextstrain/augur/pull/1017 - maybe it uses an upstream generated tree as input? Rather than an unchanging input file?

See: https://github.com/nextstrain/augur/runs/7764866231?check_suite_focus=true#step:10:18

File: augur/refine.py (hunk 2 above)
Comment (corneliusroemer) [reply augur/refine.py:15]: Silly question, it doesn't change CLI.

But FFT will now always used - and cannot be disabled via CLI.

**Comment by rneher** [comment]
this failing test is likely related to 

https://github.com/neherlab/treetime/commit/7b521242cc1e9d0aca8b3e25f77b84cc14a66498

The way missing data was handled in a haploid vcf file wasn't correct. This will result in some of the amino acids by translated as `X`.

**Comment by rneher** [comment]
these [tests](https://github.com/nextstrain/augur/pull/1017/files) specify positions that are likely supposed to be reference allele as `.` which should be interpreted as missing value according to vcf spec.

**Comment by corneliusroemer** [comment]
@rneher Do I understand correctly that now the behaviour in treetime 0.9.2 is correct - in previous versions, including 0.8.6 it was wrong. Hence we should set the new output as correct in the test?

As in - the test fails because we no longer produce an incorrect result, and the test expects the wrong result?

**Comment by corneliusroemer** [comment]
To verify that now the expected test output is correct, check the diff here: https://github.com/nextstrain/augur/commit/735db62f42f25e1007a7b483035dbd12ffaef471

The only changes impact X as expected by the fix pointed to by @rneher

So I think now we're good to go and merge this here (from my side at least).

**Comment by corneliusroemer** [comment]
Next tests fails unfortunately. Let me know if you know why: https://github.com/nextstrain/augur/runs/7766923240?check_suite_focus=true#step:11:258

Something about the tb snakemake test workflow.

**Comment by rneher** [comment]
I think you fix for the tests is wrong. I think the problem is how the vcf files are specified, not the output of the step. maybe @huddlej has an idea.

**Comment by corneliusroemer** [comment]
It could be that we need to change something in translate - not sure.

But the new error is unrelated to my potentially wrong fix. It could have the same root cause though.

We're now getting this error in the VCF snakemake workflow with treetime-0.9.2 but not with treetime-0.8.6:

Treetime 0.8.6:
```
tests/builds/tb_drm on  master [$!] via 🅒 augur-pip-dev 
❯ augur tree --exclude-sites data/drm_sites.txt --alignment results/masked.vcf.gz --vcf-reference data/ref.fasta --output results/tree_raw.nwk --method fasttree
162 masking sites read from data/drm_sites.txt
Cannot specify model unless using IQTree. Model specification ignored.
Building a tree via:
        FastTreeMP -nt -nosupport results/informative_sites.fasta 1> results/tree_raw.nwk 2> results/tree_raw.nwk.log
        Price et al: FastTree 2 - Approximately Maximum-Likelihood Trees for Large Alignments.
        PLoS ONE 5(3): e9490. https://doi.org/10.1371/journal.pone.0009490


Building original tree took 0.06926083564758301 seconds
```

Treetime 0.9.2:
```
tests/builds/tb_drm on  upgrade-treetime-0.9 [$!] via 🅒 augur-pip-dev 
❯ augur tree --exclude-sites data/drm_sites.txt --alignment results/masked.vcf.gz --vcf-reference data/ref.fasta --output results/tree_raw.nwk --method fasttree
162 masking sites read from data/drm_sites.txt
Traceback (most recent call last):
  File "/Users/cr/code/augur/augur/__init__.py", line 65, in run
    return args.__command__.run(args)
  File "/Users/cr/code/augur/augur/tree.py", line 453, in run
    variable_fasta = write_out_informative_fasta(compress_seq, args.alignment, stripFile=args.exclude_sites)
  File "/Users/cr/code/augur/augur/tree.py", line 333, in write_out_informative_fasta
    align = np.rot90(sites)
  File "<__array_function__ internals>", line 180, in rot90
  File "/usr/local/Caskroom/mambaforge/base/envs/augur-pip-dev/lib/python3.10/site-packages/numpy/lib/function_base.py", line 223, in rot90
    raise ValueError("Axes must be different.")
ValueError: Axes must be different.


An error occurred (see above) that has not been properly handled by Augur.
To report this, please open a new issue including the original command and the error above:
    <https://github.com/nextstrain/augur/issues/new/choose>
```

**Comment by rneher** [comment]
the second error is also due to a non-standard vcf. the main tb workflow runs fine, but the drug resistance mutation test `tb_drm` fails because the compressed alignment ends up being empty.

**Comment by huddlej** [comment]
Yeah, there are two bugs:

1. The VCF for the TB DRM test doesn't have any valid sites after filtering out DRM sites, so this input file needs to be replaced with a valid input. Alternately, we could try to find a minimal set of DRM sites that would allow us to test the sequence-traits command, etc. without producing an empty alignment.
2. The `write_out_informative_fasta` function doesn't check for an empty alignment. An empty alignment should cause this function to raise an exception with a meaningful error message that we catch in the `run` function where we print the error and then exit with a non-zero code.

The second bug is easy to fix. The first one either requires more knowledge of that TB dataset or a willingness to replace the TB dataset with something more generic (or better documented). @emmahodcroft might have a preference about this...

**Comment by corneliusroemer** [comment]
If I understand it correctly, neither of these bugs outlined by @huddlej are blocking for this PR. We could just comment out the tests for now until they are fixed?

Do I understand it correctly that the only reason these tests passed in the past is that tree time did not behave according to the standard. And now that this is fixed in 0.9.2 we see the problem with the test?

**Comment by rneher** [comment]
I fixed the drm vcf file (replaced `.` genotype by `0/0`) and added a check for non-empty alignment.

**Comment by rneher** [comment]
Tests pass now!

**Comment by huddlej** [comment]
Thanks, @rneher!

> If I understand it correctly, neither of these bugs outlined by @huddlej are blocking for this PR. We could just comment out the tests for now until they are fixed?
> 
> Do I understand it correctly that the only reason these tests passed in the past is that tree time did not behave according to the standard. And now that this is fixed in 0.9.2 we see the problem with the test?

I totally see the temptation to comment out breaking tests, especially when they don't appear to relate to the PR directly. In this case, though, those broken tests capture an important change TreeTime's VCF reader implementation that is directly related to this PR (the "upgrading TreeTime" bit, not the FFT bit). For similar cases in the future, we should consider those broken tests as blockers for the PR or have a good reason to exclude the tests completely.

We'd also benefit from revisiting our test suite every so often to make sure it captures what we want. It's easy for useless tests to accrue over time (or for useful tests to become irrelevant) and those tests need tending like the rest of the "main" code. For example, I'd like to replace these Snakemake tests called by the runner.sh script with Cram equivalents that are more modular, easier to run in parallel, and less redundant with other existing functional tests.

**Comment by rneher** [comment]
Agree regarding the tests. We should make an effort to find tests remain relevant. but that is a different conversation.

**Comment by rneher** [comment]
@corneliusroemer I adapted the version requirement as per your suggestion.

**Pull request merged**